infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,28 +73,40 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheHash.get(conflictKey);
+    }
+
+    private long firstAddress() {
+        if (cacheEntries.peek() == null)
+            return Address.NOT_FOUND;
+        return cacheEntries.peek().txVersion;
     }
 
     /**
-     * Invalidate all records up to a trim mark.
+     * Invalidate the first record with the minAddress.
+     */
+    private void invalidateFirst() {
+        ConflictTxStreamEntry entry = cacheEntries.poll();
+        cacheHash.remove(entry.conflictTxStream);
```

Comment:
The remove is not necessary and its also inefficient. `poll` will `reheapify`  and then `remove` will do a linear scan. It also doesn't evict multiple keys which belong to the same version, so if the heap looks like [k1:1, k2:1], only k1 will be evicted, but k2 cannot be used, so it also needs to be evicted.